abstract_domains/numerical/octagon.py

```diff
@@ -0,0 +1,578 @@
+        def generic_visit(self, expr, *args, **kwargs):
+            raise ValueError(
+                f"{type(self)} does not support generic visit of expressions! "
+                f"Define handling for expression {type(expr)} explicitly!")
+
+        def _to_LtE_operator(self, expr: BinaryComparisonOperation):
+            # noinspection PyPep8Naming
+            ConditionSet = OctagonDomain.SmallerEqualConditionTransformer.ConditionSet
+
+            if expr.operator == BinaryComparisonOperation.Operator.LtE:
+                return ConditionSet(
+                    expr)  # desired operator already there
+            elif expr.operator == BinaryComparisonOperation.Operator.GtE:
+                return ConditionSet(
+                    BinaryComparisonOperation(expr.typ, expr.right, BinaryComparisonOperation.Operator.LtE,
+                                              expr.left))
+            elif expr.operator == BinaryComparisonOperation.Operator.Lt:
+                return ConditionSet(
+                    BinaryComparisonOperation(expr.typ, \
+                                              BinaryArithmeticOperation(expr.typ, expr.left,
+                                                                        BinaryArithmeticOperation.Operator.Add,
+                                                                        Literal(int, '1')),
+                                              BinaryComparisonOperation.Operator.LtE, expr.right))
+            elif expr.operator == BinaryComparisonOperation.Operator.Gt:
+                return self._to_LtE_operator(
+                    BinaryComparisonOperation(expr.typ, expr.right, BinaryComparisonOperation.Operator.Lt, expr.left))
+            elif expr.operator == BinaryComparisonOperation.Operator.Eq:
+                cs1 = self._to_LtE_operator(
+                    BinaryComparisonOperation(expr.typ, expr.left, BinaryComparisonOperation.Operator.LtE, expr.right))
+                cs2 = self._to_LtE_operator(
+                    BinaryComparisonOperation(expr.typ, expr.left, BinaryComparisonOperation.Operator.GtE, expr.right))
+                return ConditionSet([cs1.conditions[0], cs2.conditions[0]], ConditionSet.Operator.MEET)
+
+            elif expr.operator == BinaryComparisonOperation.Operator.Eq:
+                cs1 = self._to_LtE_operator(
+                    BinaryComparisonOperation(expr.typ, expr.left, BinaryComparisonOperation.Operator.Lt, expr.right))
+                cs2 = self._to_LtE_operator(
+                    BinaryComparisonOperation(expr.typ, expr.left, BinaryComparisonOperation.Operator.Gt, expr.right))
+                return ConditionSet([cs1.conditions[0], cs2.conditions[0]], ConditionSet.Operator.JOIN)
+
+    class AssumeVisitor(ExpressionVisitor):
+        """Visits an expression and recursively 'assumes' the condition tree."""
+
+        # noinspection PyMethodOverriding
+        def visit(self, expr, state):
+            return super().visit(expr, state)
```

Comment:
This is also a bit ugly since we have to pass the state. Do we really need to have this as a subclass?